src/field/field3d.cxx

```diff
@@ -205,6 +205,13 @@ Field3D &Field3D::ynext(int dir) {
 }
 
 bool Field3D::requiresTwistShift(bool twist_shift_enabled) {
+  // Workaround for 3D coordinates.
+  // We need to communicate in the coordinates constructor in that
+  // case a Field3D, but coordinates isn't valid yet. As such we
+  // disable twist-shift in that case.
+  if (getCoordinates() == nullptr) {
+    return false;
+  }
```

Comment:
Annoyingly this is called inside `Mesh::wait` rather than `Mesh::communicate`, so it does require a bit more of a refactor. But I agree, having a "simpler" communicate would be a nicer solution, as well as allowing us to delete the re-implementation in `coordinates.cxx`.